File: domain/src/main/java/com/example/movie/repository/MovieRepository.java
package com.example.movie.repository;

import com.example.movie.entity.movie.Genre;
import com.example.movie.entity.movie.Movie;
import com.example.movie.repository.dto.MoviesDetailDto;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.time.LocalDateTime;
import java.util.List;

public interface MovieRepository extends JpaRepository<Movie,Long> {

    @Query("SELECT new com.example.movie.repository.dto.MoviesDetailDto(m.id, m.name, m.grade, m.releaseDate, m.thumbnail, m.runningTime, m.genre, " +
            "t.id, t.name, s.startAt, s.endAt) " +
            "FROM Movie m " +
            "JOIN Screening s ON m.id = s.movieId " +
            "JOIN Theater t ON s.theaterId = t.id " +
            "WHERE (:genre IS NULL OR m.genre = :genre) " +
            "AND (:search IS NULL OR LOWER(m.name) LIKE LOWER(CONCAT('%', :search, '%')))")
    List<MoviesDetailDto> findAll(@Param("now") LocalDateTime now,
                                         @Param("genre") Genre genre,
                                         @Param("search") String search);

}

File: movie/core/src/main/kotlin/com/example/movie/domain/movie/repository/MovieRepository.kt
package com.example.movie.domain.movie.repository

import com.example.movie.domain.movie.model.Movie
import com.example.movie.domain.screening.model.Screening
import com.example.movie.domain.screening.model.ScreeningStatus
import java.time.LocalDateTime

interface MovieRepository {
    fun findById(id: Long): Movie?
    fun findAllOrderByReleaseDateDesc(): List<Movie>
    fun findAllByStatusWithMovieAndTheater(currentTime: LocalDateTime, status: ScreeningStatus, title: String?, genreId: Long?): List<Movie>
}